Use Stringbuilder for strings

zonca opened the pull request:
As suggested by @ManasviGoyal, I started from https://github.com/ManasviGoyal/kaitai_struct_awkward_runtime/issues/29, that will make it easier to work on Enums.

I am currently testing on `animal.ksy`

Here are the `animal.h` and `animal.cpp` generated by this:

https://gist.github.com/zonca/87b18164ab7e62d7599ab2999889dbce

zonca committed:
feat: declare IndexedBuilder

zonca committed:
ignore vscode files

zonca committed:
feat: StringBuilder declaration

zonca committed:
feat: Scala StringBuilder class

zonca committed:
fix: typo in StringBuilder template definition

zonca committed:
feat: use StringBuilder instead of ListOffsetBuilder for strings

zonca commented on shared/src/main/scala/io/kaitai/struct/languages/AwkwardCompiler.scala:
```diff
@@ -723,7 +743,18 @@ class AwkwardCompiler(
+            outSrc.puts(s"auto& ${builderName}_stringbuilder = ${nameList.last}_builder.content<Field_${nameList.last}::${nameList.last + "A__Z" + idToStr(id)}>();")
+          else {
+            throw new NotImplementedError("StrType with repeat is not supported yet")
+            // builderName = "sub_" + builderName
```
@ManasviGoyal I am not sure how to handle repetition, could you help me creating a version of `animal.ksy` with a repeated string, so I can use that for testing?

ManasviGoyal replied:
I think we can leave this case as Not Implemented for now. As far as I know, it is not possible to have a repeat condition with a `str` type in a `ksy` file . It can be easily added later if we come across a case that requires it.

zonca commented on shared/src/main/scala/io/kaitai/struct/languages/AwkwardCompiler.scala:
```diff
@@ -1458,7 +1487,12 @@ class AwkwardCompiler(
+            builder.fields += cs.name.last + "A__Z" + idToStr(el.id)
+            var builderContent = checkRepeat(el.cond.repeat, StringBuilder("int64_t"))
+            builder.contents += checkOption(el, builderContent)
+          case _: BytesType =>
```
I kept `BytesType` as a separate case

ManasviGoyal commented on shared/src/main/scala/io/kaitai/struct/languages/AwkwardCompiler.scala:
```diff
@@ -744,10 +775,8 @@ class AwkwardCompiler(
-            outSrc.puts(s"""${builderName}_listoffsetbuilder.set_parameters("\\"__array__\\": \\"string\\"");""")
-            outSrc.puts(s"""${idToStr(id)}_builder.set_parameters("\\"__array__\\" : \\"char\\"");""")
-          }
+          outSrc.puts(s"""${builderName}_listoffsetbuilder.set_parameters("\\"__array__\\": \\"string\\"");""")
+          outSrc.puts(s"""${idToStr(id)}_builder.set_parameters("\\"__array__\\" : \\"char\\"");""")
```
For bytes, parameters will be `bytestring` instead of `string` and `byte` instead of `char`.

zonca replied:
is it possible that it is the default behavior and I do not need to specify that?

ManasviGoyal replied:
By default, the `parameters` are empty so it needs to be specified. I had missed it earlier for `BytesType`.

zonca committed:
feat: set parameters for bytestype

zonca commented:
I'll test on the runtime before we merge this

zonca commented:
what is strange is that I get 2 other unit tests failing, and they do not seem directly related to strings.
For example `index_option` fails with:

```
E           AssertionError: 
E           Arrays are not equal
E           
E           (shapes (1,), (1, 0) mismatch)
E            x: MaskedArray([masked], dtype='<U1')
E            y: MaskedArray([], shape=(1, 0), dtype=uint8)
```

However how is this related to `StrType`?

There is also an error in `test_scdms_v8`, but that is unintelligible.

ManasviGoyal commented:
> what is strange is that I get 2 other unit tests failing, and they do not seem directly related to strings. For example `index_option` fails with:
> 
> ```
> E           AssertionError: 
> E           Arrays are not equal
> E           
> E           (shapes (1,), (1, 0) mismatch)
> E            x: MaskedArray([masked], dtype='<U1')
> E            y: MaskedArray([], shape=(1, 0), dtype=uint8)
> ```
> 
> However how is this related to `StrType`?
> 
> There is also an error in `test_scdms_v8`, but that is unintelligible.

They both have string types so it's definitely related to the StringBuilder change. Can you share the generated cpp and header file for `index_option`?

zonca commented:
https://gist.github.com/zonca/38c719c1a59dc3f77bcc6643375a5375

ManasviGoyal commented:
> https://gist.github.com/zonca/38c719c1a59dc3f77bcc6643375a5375

Everything looks fine here. You should try printing the entire array and see if something is wrong there.

ManasviGoyal commented:
> the entire layout of the 2 versions are here: https://gist.github.com/zonca/c8ad33e3a8c92fe4546efaa1c91e7d6b
> 
> it seems like inside `zbank1`, the original compiler has:
> 
> ```
>                 <content index='0' field='0'>
>                     <ListOffsetArray len='0'>
>                         <offsets><Index dtype='int64' len='1'>[0]</Index></offsets>
>                         <content><NumpyArray dtype='uint8' len='0'>[]</NumpyArray></content>
>                     </ListOffsetArray>
>                 </content>
> ```
> 
> while the new one has:
> 
> ```
>                 <content index='0' field='0'>
>                     <ListOffsetArray len='0'>
>                         <parameter name='__array__'>'string'</parameter>
>                         <offsets><Index dtype='int64' len='1'>[0]</Index></offsets>
>                         <content><NumpyArray dtype='uint8' len='0'>
>                             <parameter name='__array__'>'char'</parameter>
>                             []
>                         </NumpyArray></content>
>                     </ListOffsetArray>
>                 </content>
> ```
> 
> It seems like the new implementation is just adding the 2 `parameter` tags, but that is changing the shape of the output array.
> 
> Another thing I do not understand, why this doesn't show the name of the 3 fields? `name` `type` and `size`, in `zbank2` they are present.

The parameters should have been there in the first case too so it's strange why there were none despite adding them. Regarding the fields, I think that is because in the raw data file, we only have data for `bank32` and not `bank16` (because we have a `if` condition in `ksy`). So the constructor for `bank16` is not called and the fields are not set. To fix this, `set_fields` step can be shifted to the step just before we call the respective class constructor. For example, `auto& bank32_recordbuilder = index_option_builder.content<Field_index_option::index_optionA__Zbank2>().content();` after this step. For indexedoption, you can either move this step and `set_fields` before the if-else or add them inside the else case too.

zonca commented:
I think there is something wrong with the file `index_option.raw`.
If I inspect it with the Kaitai web ide:
![image](https://github.com/ManasviGoyal/kaitai_struct_compiler/assets/383090/e0bd7443-c739-4bb2-a613-4a7e24e89a63)
it can only identify 1 `bank2` element, however, in the hex viewer it seems like there is another `Bank` string there, so maybe there were supposed to be 2 `bank2` elements?

zonca commented:
@ManasviGoyal replied:

> I think that doesn't matter much. In this case it is just reading the first element only but the layout should still be correct.

zonca commented:
fixed and simplified the index option test file as we discussed on the call:
![image](https://github.com/ManasviGoyal/kaitai_struct_compiler/assets/383090/e6ae420d-2d25-49cc-83da-0065a62c234a)

zonca commented:
@ManasviGoyal this is required to go with https://github.com/ManasviGoyal/kaitai_struct_awkward_runtime/pull/32
